### ISSUE TITLE
Poor dependency parsing performance in a voice assistant

### ISSUE THREAD

**Issue by wadetb** [explosion/spaCy#1256]
I'm writing a voice assistant for my home in Python, using spaCy to translate speech into intents. For example "turn on all the downstairs lights" or "ask google to tell a joke". 

The plan was to rely heavily on spaCy's dependency analysis, but the results have been poor, even with the 2.0 alpha. For example "turn on the basement light" ties light to on (as in, turn when you get to the light), and "ask google to tell a joke" ties to with tell (as in, to kill a mockingbird). 

```python
import spacy, pandas as pd
nlp = spacy.load('en_core_web_sm')
def dump(text):
    print(pd.DataFrame([
        {'0_orth': word.orth_, '1_dep': word.dep_, '2_head': word.head} 
        for word in nlp(text)]))
```
```
>>> dump("turn on the basement light")
     0_orth     1_dep   2_head
0      turn      ROOT     turn
1        on      prep     turn
2       the       det    light
3  basement  compound    light
4     light      pobj       on
```
```
>>> dump("ask google to tell a joke")
   0_orth  1_dep  2_head
0     ask   ROOT     ask
1  google   dobj     ask
2      to    aux    tell
3    tell  xcomp     ask
4       a    det    joke
5    joke   dobj    tell
```

The results are also quite inconsistent- sometimes a minor variation will change the behavior drastically.

```
>>> dump("turn on the bathroom light")
     0_orth 1_dep      2_head
0      turn  ROOT        turn
1        on  prep        turn
2       the   det    bathroom
3  bathroom  pobj          on
4     light  dobj        turn
```

It would be great if I could use my intents' syntax validation code to steer the dependency analysis toward viable results, as a human mind would do when considering different interpretations. I can imagine passing a `makes_sense` function into dependency analysis, which can force the parser to try a different interpretation by returning `False`.  

Alternately, there is the approach Google has taken with their speech recognition API - it can return _alternatives_ for the client to choose from. It would be great if spaCy could do this too, so I could just attempt alternatives in sequence until one makes sense. But with thousands of possible interpretations for a complex sentence, that might not be a general solution.

After asking about this in Gitter I was directed to look at training my own model. I spent quite a bit of time trying to work that process out, but ended up with a model that matched nothing but my training phrases. As a user of the package I feel like its notion of English should be essentially correct (and therefore not need further training), and the idea of trying to generate and tag random-ish expressions from my intents' syntax trees feels too much like throwing darts. 

I also implemented named entity matching, as in detecting all my known entities and pre-tagging them as themselves, thus resolving confusion about "bathroom lights", but that didn't help with the interpretation of "on".

Anyway, as it stands I'm stuck here with high level hacks - matching inputs sequentially instead of structurally, and intentionally matching against incorrect grammatical interpretations. Other than that I'm loving spaCy's power and ease-of-use, so thank you for making this product available, and any help you can give would be greatly appreciated.

## Info about spaCy

* **Python version:** 3.5.3
* **spaCy version:** 2.0.0a6
* **Platform:** Linux-4.9.0-3-amd64-x86_64-with-debian-9.0

**Comment by honnibal** [comment]
> After asking about this in Gitter I was directed to look at training my own model. I spent quite a bit of time trying to work that process out, but ended up with a model that matched nothing but my training phrases.

Try parsing a bunch of text with the default model, and using that in the updates as well. This prevents the "catastrophic forgetting" problem.

When you start optimising with just the sentences in your training set, you're setting the optimiser the problem, "Minimise this loss". If you then let it converge, it'll find one of the many solutions that minimises the loss on those sentences  --- which may or may not be useful otherwise. If you add the previously parsed sentences, you set the problem "Match the previous behaviour, but change behaviour on these sentences". This is closer to what you want.

> As a user of the package I feel like its notion of English should be essentially correct (and therefore not need further training)

Well...We aspire to this, and I think spaCy 2's current balance of speed, model size and accuracy across genres compares favourably to anything else out there. However, the model doesn't know you're only trying to parse voice commands, and it actually sees this type of grammar as quite unlikely, because most of its training data comes from web text.

**Comment by wadetb** [comment]
Thanks for the suggestion about training the model more effectively, I will try that. So the idea is to use its own dependency analysis to train it on some corpus of text (I suppose I'll have to find one somewhere....), while nudging it towards the correct interpretation for my context. I'll also need to adapt my intent syntax trees to synthesize random sentences to train the model completely.

In the long term, I hope the state of the art will move in the direction of integrating high level logic into the disambiguation phase of these deep neural network-based algorithms. I am sure that spaCy has the right answer already in its existing model, just not with the absolute highest weight.

**Comment by honnibal** [comment]
Yes, that's the idea. But there's an important caveat: make sure you parse the text with the *original* model. You mustn't let the model set its own gold standard itself, particularly with quick iterations. If you do, the model will be drawn into states where the solution is trivial, e.g. entirely left-branching analyses. Once the model starts predicting one of these trivial analyses, it won't be able to escape.

So you need to parse the text first and keep the objective fixed. That way the objective stays anchored by the original annotations.
